File: py/structure/set/BitMap.py
#!/usr/bin/env python
# -*- coding:gbk -*-

class BitMap:
    def __init__(self):
        self.__hash = [0x00 for _ in range(65536)]

    def isEmpty(self) -> bool:
        for value in self.__hash:
            if value:
                return False
        return True

    def add(self, value: int) -> None:
        index = value >> 3
        if index >= 0 and index < 65536:
            self.__hash[index] |= 1 << (value & 7)

    def erase(self, value: int) -> None:
        index = value >> 3
        if index >= 0 and index < 65536:
            self.__hash[index] &= ~(1 << (value & 7))

    def contains(self, value: int) -> bool:
        index = value >> 3
        if index >= 0 and index < 65536:
            return bool(self.__hash[index] & (1 << (value & 7)))
        return False

    def clear(self) -> None:
        self.__hash = [0x00 for _ in range(65536)]

File: cpp/structure/set/BitMap.cpp
#include <iostream>
#include <vector>
using namespace std;

class BitMap {
public:
    BitMap() : hash(vector<char>(65536, 0x00)) {}

    bool isEmpty() {
        for (char value : hash) {
            if (value) {
                return false;
            }
        }
        return true;
    }

    void add(int value) {
        int index = value >> 3;
        if (index >=0 && index < 65536) {
            hash[index] |= 1 << (value & 7);
        }
    }

    void erase(int value) {
        int index = value >> 3;
        if (index >=0 && index < 65536) {
            hash[index] &= ~(1 << (value & 7));
        }
    }

    bool contains(int value) {
        int index = value >> 3;
        if (index >=0 && index < 65536) {
            return 0 != (hash[index] & (1 << (value & 7)));
        }
        return false;
    }

    void clear() {
        hash.assign(65536, 0x00);
    }

private:
    vector<char> hash;
};
